### PR TITLE
Prevent modifying and/or deleting reserved statuses

### LINKED ISSUE #6291: Important Statuses can be deleted and modified
## This Bug Report affects these Traffic Control components:
- Traffic Ops

## Current behavior:
The Statuses named `OFFLINE`, `ONLINE`, `REPORTED`, `PRE_PROD`, and `ADMIN_DOWN` are handled specially by different parts of Traffic Control, and different parts of the code even go so far as to assume that `OFFLINE`, `ONLINE`, and `REPORTED` exist, but in fact there is no guarantee that any of the five exist at any given time.

That's because you can rename them (and less importantly change their descriptions):
```http
GET /api/4.0/statuses?id=1 HTTP/1.1
Host: localhost:6443
Cookie: mojolicious=...
Content-Length: 0


HTTP/1.1 200 OK
Content-Type: application/json
Set-Cookie: mojolicious=...
Transfer-Encoding: chunked

{ "response": [{
	"description": "Server is Offline. Not active in any configuration.",
	"id": 1,
	"lastUpdated": "2021-10-14 08:04:39+00",
	"name": "OFFLINE"
}]}
```
```http
PUT /api/3.0/statuses/1 HTTP/1.1
Cookie: mojolicious=...
Content-Length: 88
Content-Type: application/json

{"name": "Not OFFLINE anymore", "description": "an unhelpful or inaccurate description"}


HTTP/1.1 200 OK
Content-Type: application/json
Permissions-Policy: interest-cohort=()
Set-Cookie: mojolicious=...
Content-Length: 188

{
	"alerts": [
		{
			"text": "status was updated.",
			"level": "success"
		}
	],
	"response": {
		"description": "an unhelpful or inaccurate description",
		"id": 1,
		"lastUpdated": "2021-10-15 08:41:42+00",
		"name": "Not OFFLINE anymore"
	}
}
```

and even delete them:
```http
DELETE /api/4.0/statuses/1 HTTP/1.1
Host: localhost:6443
Cookie: mojolicious=...
Content-Length: 0


HTTP/1.1 200 OK
Content-Type: application/json
Transfer-Encoding: chunked
Set-Cookie: mojolicious=...

{ "alerts": [{
	"text": "status was deleted.",
	"level": "success"
}]}
```

## Expected behavior:
You shouldn't be able to get rid of things that we assume should exist. That means either disallowing those operations on certain of those objects, or disallowing those operations entirely. I've always been hopeful that we could do away with custom statuses, eliminating a server field, two endpoints, hundreds of lines of code, two database tables, and a lot of confusion and edge cases like this. However, that behavior is depended upon by many, so we can't get rid of it without at least an alternative (like the one proposed by #4819), so if nothing else we should ensure that all valid (i.e. only manipulated through the API) Traffic Ops server states contain the Statuses that ATC requires to function. At a minimum that would be `REPORTED`, but to avoid unexpected behavior when creating new Statuses with reserved names we should also do the same with the others mentioned above.

### PR DESCRIPTION
**Pull request by srijeet0406**
This PR closes #6291 

<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

Run TO and TP locally.
Try to modify/ delete one of the following statuses. You should get an error while doing so.
Reserved status names:
`PRE_PROD, ONLINE, OFFLINE, ADMIN_DOWN, REPORTED`
## If this is a bugfix, which Traffic Control versions contained the bug?

- master

## PR submission checklist
- [x] This PR has tests 
- [ ] This PR has documentation 
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

**Comment by mitchell852** [comment]
just curious. is "PRE_PROD" a reserved name? i.e. is there code that looks specifically for PRE_PROD and performs certain actions based on that status? is this status seeded? i was kind of under the impression Comcast added this status to support an internal process.

**Comment by srijeet0406** [comment]
> just curious. is "PRE_PROD" a reserved status? i.e. is there code that looks specifically for PRE_PROD and performs certain actions based on that status? is this status seeded? i was kind of under the impression Comcast added this status to support an internal process.

I added it in there for two reasons:
1. The issue mentions it.
2. Our seeds file adds this status as part of the seeding process, so I assumed it was reserved.
That said, I can remove it if we think that it shouldn't be reserved. Not sure if/ when it got to be a reserved status.

**Comment by ocket8888** [comment]
> just curious. is "PRE_PROD" a reserved status? i.e. is there code that looks specifically for PRE_PROD and performs certain actions based on that status? is this status seeded? i was kind of under the impression Comcast added this status to support an internal process.

Comcast may have done that, but it's seeded into TODB on installation. There is no code in ATC that checks for and acts explicitly on that status. It has no non-semantic meaning.

**Comment by mitchell852** [comment]
> Comcast may have done that, but it's seeded into everyone's TODB on installation

ok well is guess since it's "seeded" it should be immutable

File: traffic_ops/traffic_ops_golang/status/statuses.go [hunk 1]
```diff
@@ -40,6 +40,14 @@ type TOStatus struct {
...
 }
 
+var statusNameMap = map[tc.CacheStatus]bool{
+	tc.CacheStatusAdminDown: true,
+	tc.CacheStatusOnline:    true,
+	tc.CacheStatusOffline:   true,
+	tc.CacheStatusReported:  true,
+	tc.CacheStatusPreProd:   true,
+}
+
```
Comment (ocket8888) [inline traffic_ops/traffic_ops_golang/status/statuses.go:50]: This map is not used

File: traffic_ops/testing/api/v4/statuses_test.go [hunk 2]
```diff
@@ -142,13 +145,26 @@ func GetTestStatusesIMS(t *testing.T) {
...
-		if err != nil {
-			t.Errorf("could not create Status: %v - alerts: %+v", err, resp.Alerts)
+		if status.Name != nil {
+			if _, ok := statusNameMap[*status.Name]; !ok {
+				resp, _, err := TOSession.CreateStatus(status, client.NewRequestOptions())
+				t.Log("Response: ", resp)
+				if err != nil {
+					t.Errorf("could not create Status: %v - alerts: %+v", err, resp.Alerts)
+				}
+			}
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/statuses_test.go:165]: You wouldn't need to build this map and do these checks if you just remove standard statuses from the testing fixture data.

File: lib/go-tc/statuses.go [hunk 3]
```diff
@@ -70,3 +70,19 @@ type StatusNullable struct {
 	LastUpdated *TimeNoMod `json:"lastUpdated" db:"last_updated"`
 	Name        *string    `json:"name" db:"name"`
 }
+
+func IsReservedStatus(status string) bool {
```
Comment (ocket8888) [inline lib/go-tc/statuses.go:74]: This function could use a GoDoc.

**Review by ocket8888 — COMMENTED** [review]
The changes to the tests are basically the same in all four of the places you're making them, so I just put the comment I had on v4 to avoid repeating myself a bunch of times.

Also, though, now that any cache status that t3c could possibly care about is guaranteed to always exist, they shouldn't need to have any fixture data or testing functions at all. Can probably even be removed from their `TCObj`/`WithObj` stuff.

File: cache-config/testing/ort-tests/tcdata/statuses.go [hunk 4]
```diff
@@ -17,30 +17,14 @@ package tcdata
```
Comment (ocket8888) [inline cache-config/testing/ort-tests/tcdata/statuses.go:17]: Are the two functions in this file ever called anymore with all of the removals you did? I wouldn't think so, but I could be wrong.

Comment (srijeet0406) [reply cache-config/testing/ort-tests/tcdata/statuses.go:17]: You're right, I've removed the file now.

**Review by ocket8888 — APPROVED** [review]

**Pull request merged**